File: datasets/python_source-code/Multiple-C2/organic_7.py
"""
0/1 Knapsack (DP) - Organic Submission #7
Accepts items as a list of (name, weight, value) tuples and iterates
capacity as the outer loop, items as the inner loop (column-major order).
"""


def knapsack(items, capacity):
    n = len(items)
    dp = [[0] * (n + 1) for _ in range(capacity + 1)]

    for cap in range(capacity + 1):
        for idx in range(1, n + 1):
            _, weight, value = items[idx - 1]
            if weight <= cap:
                dp[cap][idx] = max(dp[cap][idx - 1], value + dp[cap - weight][idx - 1])
            else:
                dp[cap][idx] = dp[cap][idx - 1]

    return dp[capacity][n]


if __name__ == "__main__":
    items = [
        ("gold_bar", 3, 60),
        ("silver_bar", 2, 40),
        ("gem", 4, 70),
        ("statue", 5, 90),
    ]
    print(knapsack(items, 8))

File: server/datasets/java_source-code/Multiple-C2/organic_7.java
/*
 * 0/1 Knapsack (DP) - Organic Submission #7
 * Column-major DP order with capacity as outer loop and items as inner loop.
 */

class Solution {
    public static int knapsack(int[] weights, int[] values, int capacity) {
        int n = weights.length;
        int[][] dp = new int[capacity + 1][n + 1];

        for (int cap = 0; cap <= capacity; cap++) {
            for (int idx = 1; idx <= n; idx++) {
                int weight = weights[idx - 1];
                int value = values[idx - 1];
                if (weight <= cap) {
                    dp[cap][idx] = Math.max(dp[cap][idx - 1], value + dp[cap - weight][idx - 1]);
                } else {
                    dp[cap][idx] = dp[cap][idx - 1];
                }
            }
        }
        return dp[capacity][n];
    }

    public static void main(String[] args) {
        int[] weights = {3, 2, 4, 5};
        int[] values = {60, 40, 70, 90};
        System.out.println(knapsack(weights, values, 8));
    }
}
